### ISSUE TITLE
test_toy.py produces "Wrong homopolymer" error on aarch64 machine

### ISSUE THREAD

**Issue by rungitta** [fenderglass/ABruijn#398]
I get this error when running `test_toy.py` on an aarch64 machine. I tried swapping to the Makefile here:
https://github.com/zovoilis-lab/Flye_ppcle64/blob/flye/lib/minimap2/Makefile
But that didn't help. Any ideas how to fix this?

```bash
[2021-06-11 16:40:48] INFO: >>>STAGE: polishing
[2021-06-11 16:40:48] INFO: Polishing genome (1/1)
[2021-06-11 16:40:48] INFO: Running minimap2
[2021-06-11 16:40:50] INFO: Separating alignment into bubbles
[2021-06-11 16:41:02] INFO: Alignment error rate: 0.009609
[2021-06-11 16:41:02] INFO: Correcting bubbles
terminate called after throwing an instance of 'std::runtime_error'
  what():  Wrong homopolymer
[2021-06-11 16:41:03] ERROR: Command '['flye-modules', 'polisher', '--bubbles', '/home/palrosen/Flye/flye_toy_test/40-polishing/bubbles_1.fasta', '--subs-mat', '/home/palrosen/.local/lib/python3.6/site-packages/flye/config/bin_cfg/pacbio_chm13_substitutions.mat', '--hopo-mat', '/home/palrosen/.local/lib/python3.6/site-packages/flye/config/bin_cfg/pacbio_chm13_homopolymers.mat', '--out', '/home/palrosen/Flye/flye_toy_test/40-polishing/consensus_1.fasta', '--threads', '8']' died with <Signals.SIGABRT: 6>.
[2021-06-11 16:41:03] ERROR: Pipeline aborted
Traceback (most recent call last):
  File "flye/tests/test_toy.py", line 41, in <module>
    sys.exit(main())
  File "flye/tests/test_toy.py", line 36, in main
    test_toy()
  File "flye/tests/test_toy.py", line 30, in test_toy
    "-o", out_dir, "-t", "8", "-m", "1000"])
  File "/usr/lib64/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['flye', '--pacbio-corr', '/home/palrosen/Flye/flye/tests/data/ecoli_500kb_reads_hifi.fastq.gz', '-g', '500k', '-o', 'flye_toy_test', '-t', '8', '-m', '1000']' returned non-zero exit status 1.
```

**Comment by mikolmogorov** [comment]
@rungitta is it possible that you are using outdated binaries? Try a clean local build (after git clone) and invoke using `bin/flye` - does it give you the same error?

**Comment by rungitta** [comment]
I fixed it with this patch:

```bash
diff -ru Flye.orig/src/Makefile Flye/src/Makefile
--- Flye.orig/src/Makefile      2021-06-14 16:02:46.709190000 +0200
+++ Flye/src/Makefile   2021-06-14 16:03:31.146916000 +0200
@@ -1,6 +1,6 @@
 .PHONY: all clean debug profile

-CXXFLAGS += -Wall -Wextra -pthread -std=c++11 -g
+CXXFLAGS += -Wall -Wextra -pthread -std=c++11 -g -fsigned-char
 LDFLAGS += -pthread -std=c++11 -rdynamic

 MODULES_BIN := ${BIN_DIR}/flye-modules
```

**Comment by mikolmogorov** [comment]
@rungitta interesting! I looks like there might be a potential integer overflow somewhere - I will double check.

**Comment by rungitta** [comment]
This is the error we saw:

polishing/subs_matrix.cpp: In constructor 'HopoMatrix::State::State(const string&, size_t, size_t)':
polishing/subs_matrix.cpp:157:16: warning: comparison is always false due to limited range of data type [-Wtype-limits]
    if (runNucl == -1) runNucl = str[i];
        ~~~~~~~~^~~~~

My collegue suggested this patch (the other one I posted is a work around):

diff --git a/src/polishing/subs_matrix.cpp b/src/polishing/subs_matrix.cpp
index 7769734a..2b289f26 100644
--- a/src/polishing/subs_matrix.cpp
+++ b/src/polishing/subs_matrix.cpp
@@ -149,12 +149,16 @@ HopoMatrix::State::State(const std::string& str, size_t start, size_t end)

        //std::cerr << str.substr(start, end - start) << std::endl;
        size_t runLength = 0;
-       char runNucl = -1;
+       char runNucl;
+       int runNuclSet = 0;
        for (size_t i = start; i < end; ++i)
        {
                if (str[i] != '-')
                {
-                       if (runNucl == -1) runNucl = str[i];
+                       if (runNuclSet == 0) {
+                               runNuclSet = 1;
+                               runNucl = str[i];
+                       }
                        if (str[i] != runNucl) throw std::runtime_error("Wrong homopolymer");
                        ++runLength;
                }


-- 
Axel Rosén, Head Engineer
Research Infrastructure Services Group, USIT
University of Oslo, Norway

**Comment by mikolmogorov** [comment]
Fixed in 96afa77
